### PR TITLE
feat(attention): add causal + bidirectional-ranges batch prefill

### PR DESCRIPTION
**Pull request by lesj0610**
## 📌 Description

FA2 supports custom attention variants, but there is currently no packaged batch-prefill path for sequences that are causal except for spans that attend in both directions. Callers either have to materialize a dense custom mask or use a different backend.

The dense mask is `O(qo_len * kv_len)`. For example, 16384 scheduled query tokens against a 262144-token context require 4 GiB just for a boolean mask.

This PR adds an FA2 paged-prefill wrapper that computes the mask directly from a compact `int32 [total_q, 2]` range tensor:

```text
(causal AND causal_window) OR (in_range AND range_window)
```

Each query row contains an inclusive `[start, end]` span in request-local KV positions. `(-1, -1)` means that the query has no bidirectional span and remains purely causal. At 16384 query tokens, this tensor is 128 KiB regardless of context length.

The implementation uses the existing `variant_owns_mask` path in `BatchPrefillWithPagedKVCacheWrapper`; it does not change the default prefill dispatch.

The public API added here is:

| Symbol | Exported from |
|---|---|
| `BatchPrefillWithCausalBidirectionalRangesWrapper` | `flashinfer`, `flashinfer.attention` |
| `gen_batch_prefill_bidirectional_ranges_module` | `flashinfer.jit.attention` |

The wrapper and generator share one internal spec for the URI and JIT tensor and scalar arguments.

Two kernel details are worth calling out:

- The variant sets `window_left = kv_len` to keep FA2's normal sliding-window tile pruning from discarding keys that may still belong to a bidirectional range. The causal and range windows are applied inside the variant instead.
- Padded `CTA_TILE_Q` lanes are clamped to the last real query row before indexing the range tensor.

The wrapper is FA2-only because this path relies on a custom `LogitsMask` being evaluated for every KV tile. Its JIT specialization is fixed in the constructor; `plan()` and `workspace_size()` inherit omitted dtype values from that specialization and reject explicit mismatches. Options that conflict with the variant-owned mask are rejected rather than silently ignored.

Packed NVFP4 KV caches use the existing `maybe_k_cache_sf` and `maybe_v_cache_sf` JIT tensor conventions.

## 🔍 Related Issues

None.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

```bash
pytest tests/attention/test_bidirectional_ranges_attention.py
# 39 passed, 2 warnings, on one SM80 device
```

The targeted tests cover:

- dense-reference correctness with multiple requests, GQA and shuffled page IDs;
- independent causal-window and bidirectional-range behavior;
- partial query tiles and padded lanes;
- packed NVFP4 KV, separate output dtype and non-unit Q/K/V scales;
- CUDA graph replay while updating the same range buffer;
- constructor/plan specialization mismatches and unsupported mask options.

Pre-commit hooks for the changed files also pass.

## Reviewer Notes

This is currently JIT-only. Unlike the existing attention-sink AOT entry, it does not have a small fixed set of dtype and head-dimension combinations to precompile.

Testing was limited to SM80. I did not run this on SM90 or SM100, and the FP8 paths were not exercised.

This wrapper is separate from `prefix_len_ptr`, which selects `MaskMode.MULTIITEMSCORING`; the two mask modes cannot be combined.

## Summary by CodeRabbit

- **New Features**
  - Added bidirectional range attention for batch prefill, combining causal masking with per-query key spans.
  - Exposed the new attention wrapper through the public API.
  - Added support for NVFP4 key-value caches and log-sum-exp results.
  - Added validation for range inputs, attention settings, data types, and configuration compatibility.
  - Added tensor-valued query, key, and value scaling support for paged prefill.

- **Documentation**
  - Documented causal and bidirectional ranges prefill functionality.

### CODE REVIEW

**Comment by lesj0610** [comment]
@qsang-nv, could you review this one? It adds an FA2 prefill variant for causal attention with per-query bidirectional ranges. The two parts I’m least sure about are the mask semantics and where I drew the boundary between the public wrapper and the JIT API.

File: flashinfer/prefill.py [hunk 1]
```diff
@@ -2788,6 +2788,7 @@ def run(
         q: torch.Tensor,
         paged_kv_cache: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]],
         *args,
+        q_scale: Optional[float] = None,
```
[the remark this answers is not in the document]
Comment (lesj0610) [reply flashinfer/prefill.py]: Fixed in a4cd0bc. The implementation takes `Union[float, torch.Tensor]` for all three calibration scales, so the overloads were wrong to narrow them.

`k_scale` and `v_scale` carried the same narrowing before this PR and `q_scale` inherited it when I added it, so I widened all three rather than leave two knowingly-wrong annotations next to the corrected one. Full mypy and the rest of pre-commit pass.

**Comment by qsang-nv** [comment]
Reviewed the variant and the wrapper against current `main`. The `window_left = kv_len` trick is sound — `kv_start_idx` collapses to 0 since `qo_tile_idx * CTA_TILE_Q / group_size < qo_len`, and the split-KV `min(kv_len_safe, window_left + CTA_TILE_Q)` opens to the full `kv_len` — and the `prepare_jit_additional_args` ordering is correct for both the NVFP4 and non-NVFP4 specs. Five findings.

**1. `causal_window_left` is off by one against the library's `window_left`, and `0` is inverted.**

`StandardAttention` in `variants.cuh` keeps a key when `kv_idx + qo_len + window_left >= kv_len + qo_idx`, i.e. `q_abs - kv <= window_left`: `N` keeps `N + 1` keys and `0` keeps only the diagonal. The new variant uses `(q_abs - kv) < sw_left` with `sw_left <= 0` disabling, so `causal_window_left = N` equals `window_left = N - 1`, and `causal_window_left = 0` means unbounded. `plan()` rejects `window_left` and redirects callers to `run(causal_window_left=...)`, which is exactly the path that hits this. The test reference encodes the new convention (`dist < causal_window_left`), so it cannot surface the difference.

**2. `run_return_lse` bypasses the subclass; `forward_return_lse` cannot work at all.**

`run_return_lse = functools.partialmethod(run, return_lse=True)` captures the parent's function object rather than doing a name lookup. This is the first subclass to override `run` — `BatchAttentionWithAttentionSinkWrapper` only overrides `__init__` — so `wrapper.run_return_lse(q, kv_cache, ranges, ...)` goes straight to the parent and skips the override-only checks: `bidirectional_ranges` dtype / shape / row-count / device / contiguity, and the tensor-scale rejections. (`q.dtype` is still covered by the parent's `_check_cached_qkv_data_type`.) A non-contiguous range tensor reaches the kernel with the wrong row stride.

`forward_return_lse` has no parameter for the ranges, so it raises `JIT module declares additional scalar 'causal_window_left', ...` out of `prepare_jit_additional_args` — after already assigning `self._causal`, `self._window_left`, `self._logits_soft_cap` and `self._sm_scale`, the plan state this wrapper otherwise refuses to let a caller set. Neither inherited entry point is documented as unsupported here.

**3. FP8 `q_data_type` fails at nvcc; FP8 `o_data_type` bypasses the existing guard.**

`gen_batch_prefill_module` guards fa2 with `assert not fp8_enabled, "fp8 tensor core is not supported in fa2 backend"` and `assert dtype_o not in [torch.float8_e4m3fn, torch.float8_e5m2], "FP8 output is not supported in fa2/fa3 backends yet"`. `gen_customize_batch_prefill_module`, which the new generator calls directly, has neither, and the constructor canonicalizes whatever dtype it is handed. `q_data_type=torch.float8_e4m3fn` therefore renders sources and fails on `static_assert(sizeof(DTypeQ) == 2)`; FP8 `o_data_type` reaches the generator with the standard guard simply absent — I did not trace where it lands. `gen_batch_prefill_attention_sink_module` shares this gap, but the constructor and the public generator are two new entry points into it, and the constructor already refuses `pos_encoding_mode != "NONE"` on the identical "would not compile" grounds.

**4. The trace identity is a plain causal prefill, and Trace Apply cannot tell the two apart.**

The override is undecorated, but `super().run()` is `@flashinfer_api(trace=gqa_paged_prefill_trace)`, whose auto-dump fires on every call under `FLASHINFER_TRACE_DUMP=1`. That template declares only `q / k_cache / v_cache / qo_indptr / kv_indptr / kv_indices / sm_scale / return_lse` — no ranges, no windows — its `reference` is a pure causal implementation, and it is tagged `status:verified`. `wrapper.run.fi_trace` also does not exist, since the subclass `run` is undecorated.

Trace Apply patches by `fi_api` = `flashinfer.prefill.BatchPrefillWithPagedKVCacheWrapper.run`, installed on the class, so `super().run()` reaches the patched attribute through the MRO. Its lookup key is `template.definition_name(axes)` — `name_prefix` plus the const axes `num_qo_heads`, `num_kv_heads`, `head_dim`, `page_size` — none of which distinguish the mask. Read statically, a solution registered from an ordinary causal prefill at the same head counts, head dim and page size matches a call to this wrapper exactly, and a matched solution is used rather than falling back: silently wrong for any call whose spans are not all sentinel.

`BatchAttentionWithAttentionSinkWrapper` has the same shape with `sinks`, so the gap predates this PR. What is new is that this wrapper exists to apply a mask the template does not describe, is exported from `flashinfer/__init__.py`, and as it stands has no trace identity distinct from the parent's causal template and nothing excluding it from Trace Apply dispatch.

**5. Testing gap: no case reaches split-KV, and the only hardware data is SM80.**

Every case uses `kv_len <= 96`. The second consumer of `window_left = kv_len` is the split-KV write-out, `num_kv_chunks = ceil_div(min(kv_len_safe, window_left + CTA_TILE_Q), kv_chunk_size)`, which no test reaches; `fixed_split_size`, `disable_split_kv`, `max_token_per_sequence` and `max_sequence_kv` are forwarded by `plan()` with no coverage. That path does not depend on the context length in the motivation: `plan()` already forwards `fixed_split_size`, and `scheduler.cuh` assigns `kv_chunk_size = fixed_split_size` whenever it is positive and `disable_split_kv` is false. Separately, `CTA_TILE_KV` is selected from `max_smem_per_sm` / `max_smem_per_block_optin` and so differs on SM90 and SM100, while the `CTA_TILE_Q` padding-lane clamp is precisely a tile-size-dependent path.

**Comment by lesj0610** [comment]
@qsang-nv Thanks for the careful review. I fixed the five issues in `de4267ed`, with a small typing-only follow-up in `6383e2c3`.

The window behavior now matches the standard `window_left` semantics, including `0` as diagonal-only. The return-LSE and deprecated forward paths now go through the ranges-aware wrapper as well. I also added early checks for unsupported FP8 query and output types, kept this wrapper out of the parent's causal trace and Trace Apply path, and added split-KV coverage that compares split and unsplit output and LSE.

While adding the planning coverage, I found that `max_sequence_kv` currently fails inside the parent planner. This wrapper now rejects it with a clear error instead. Runtime head-count and head-dimension mismatches are also checked before dispatch.

The focused suite passes all 72 tests on SM80, and the pre-commit, documentation, and public API checks pass. I could not validate SM90 or SM100.

**Comment by qsang-nv** [comment]
The FP8, tracing, split-KV and `run` / `run_return_lse` / `forward` fixes check out — the Trace Apply exclusion is verified the right way, with a tripwire on the parent attribute plus a precondition test that the patch was really installed, and the split-KV cases read `split_kv` from `_plan_info` rather than inferring it from the KV length. Two issues remain.

**1. `begin_forward` is the same defect as `run_return_lse`.**

The parent has exactly two class-body aliases that capture a parent function object: `run_return_lse = functools.partialmethod(run, return_lse=True)` and `begin_forward = plan`. This revision rebinds the first; the second still resolves to the parent's `plan`, so `wrapper.begin_forward(...)` skips `_check_specialization`, `_reject_max_sequence_kv` and every option refusal in the override.

- `max_sequence_kv` reaches the parent planner again and dies with the `UnboundLocalError` this revision added a named error for.
- The parent's `q_data_type` default is `"float16"`, so a bf16 wrapper planned this way caches float16 and later fails in `_check_cached_qkv_data_type` — the subclass check passes first, since it compares `q.dtype` against `_spec` rather than the cached value, so the error names a dtype the caller never asked for.
- `window_left` is the silent one. `PrefillSplitQOKVIndptr` works in pages: for `window_left >= 0` each request is truncated to `min(ceil_div(window_left + cta_tile_q, page_size), kv_len_pages)`, and `num_chunks_kv`, `merge_indptr` and `o_indptr` all come from that truncated count (`kv_chunk_size` is converted to tokens only on the way out). The variant pins `window_left = kv_len`, so the kernel computes `num_kv_chunks = ceil_div(kv_len, kv_chunk_size)` over the full token length. On the split path the host then emits only the truncated number of `kv_tile_idx` work items, so no CTA visits the tail of the KV, and the kernel strides its partial writes by `num_kv_chunks * o_stride_n` while the host reserved `qo_len * num_chunks_kv` rows for that request, so the writes run past it.

`head_dim_qk` / `head_dim_vo` are the same shape of problem: `PrefillPlanImpl` reserves `tmp_v` as `num_qo_heads * padded_batch_size * cta_tile_q * head_dim_vo * sizeof(float)` from the plan-time value and records its base in `plan_info.v_offset`, while the kernel writes rows of the compile-time `HEAD_DIM_VO`. `_check_runtime_shapes` compares the tensors against `_spec`, not against what the planner was given.

**2. `range_window_left`'s boundary is undocumented and differs from the parameter above it.**

The `max(range_window_left, 0)` normalization reads as a deliberate separate convention, so this is about what the contract states:

- The positive boundary is exclusive — `(q_abs - kv) < range_window` — and the docstring gives only the off value, nothing about the edge.
- `causal_window_left`, directly above it, now promises "the same meaning as `window_left` everywhere else in FlashInfer ... no conversion is needed". The two land on different edges:

| | kept backward distances | `0` means |
|---|---|---|
| `causal_window_left = 8` | `0..8` (9 keys) | diagonal only |
| `range_window_left = 8` | `0..7` (8 keys) | unbounded |

  Negative distances are never clamped, so the forward side of a span is unaffected; the table is about the backward edge both names refer to.

- No boundary test is independent of the implementation. `causal_window_left` gained three; `range_window_left` has only `test_bidirectional_ranges_window_semantics`, which asserts that 24 differs from unclamped rather than where the edge falls, and `_dense_reference` uses the same `< range_window_left` predicate as the kernel.

**Comment by lesj0610** [comment]
@qsang-nv Thanks, both were real. `begin_forward` is rebound to this wrapper's `plan`, so the specialization and option checks run through it too, and a BF16 plan and run are covered.

I kept `range_window_left`'s count-based semantics rather than changing the kernel: `N` keeps backward distances `0..N-1`, and nonpositive values leave the range unclamped. The docstring says how that differs from `causal_window_left`, and the new boundary test fails if the edge moves by one key.

The focused suite passes 78 tests and the full pre-commit run is clean. Still SM80 only.

**Review by qsang-nv — APPROVED** [review]
